From: Openkylin Developers <packaging@lists.openkylin.top>
Date: Thu, 8 Sep 2022 02:10:22 +0000
Subject: Fix incompatibility between <stdatomic.h> and <atomic>

Bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60932
Bug: https://reviews.llvm.org/D45470

This 2 headers combined will cause errors for both GCC and Clang. This patch
makes sure only one of them is present at any time.
---
 libcutils/include/cutils/atomic.h | 16 ++++++++++++++++
 libcutils/include/cutils/trace.h  |  7 +++++++
 liblog/logger.h                   |  6 ++++++
 3 files changed, 29 insertions(+)

diff --git a/libcutils/include/cutils/atomic.h b/libcutils/include/cutils/atomic.h
index 0c88bfe..43cc5b3 100644
--- a/libcutils/include/cutils/atomic.h
+++ b/libcutils/include/cutils/atomic.h
@@ -19,7 +19,23 @@
 
 #include <stdint.h>
 #include <sys/types.h>
+#ifdef __cplusplus
+#include <atomic>
+using std::atomic_compare_exchange_strong_explicit;
+using std::atomic_fetch_add_explicit;
+using std::atomic_fetch_or_explicit;
+using std::atomic_fetch_sub_explicit;
+using std::atomic_int_least32_t;
+using std::atomic_load_explicit;
+using std::atomic_store_explicit;
+using std::atomic_thread_fence;
+using std::memory_order::memory_order_acquire;
+using std::memory_order::memory_order_relaxed;
+using std::memory_order::memory_order_release;
+using std::memory_order::memory_order_seq_cst;
+#else
 #include <stdatomic.h>
+#endif
 
 #ifndef ANDROID_ATOMIC_INLINE
 #define ANDROID_ATOMIC_INLINE static inline
diff --git a/libcutils/include/cutils/trace.h b/libcutils/include/cutils/trace.h
index fcbdc9b..4d4edb4 100644
--- a/libcutils/include/cutils/trace.h
+++ b/libcutils/include/cutils/trace.h
@@ -18,7 +18,14 @@
 #define _LIBS_CUTILS_TRACE_H
 
 #include <inttypes.h>
+#ifdef __cplusplus
+#include <atomic>
+using std::atomic_bool;
+using std::atomic_load_explicit;
+using std::memory_order_acquire;
+#else
 #include <stdatomic.h>
+#endif
 #include <stdbool.h>
 #include <stdint.h>
 #include <stdio.h>
diff --git a/liblog/logger.h b/liblog/logger.h
index 246b33c..afe96e3 100644
--- a/liblog/logger.h
+++ b/liblog/logger.h
@@ -17,7 +17,13 @@
 #ifndef _LIBLOG_LOGGER_H__
 #define _LIBLOG_LOGGER_H__
 
+#ifdef __cplusplus
+#include <atomic>
+using std::atomic_int;
+using std::atomic_uintptr_t;
+#else
 #include <stdatomic.h>
+#endif
 #include <stdbool.h>
 
 #include <cutils/list.h>
